File: Partition_Equal_Subset_Sum/solution1-2.java
class Solution {
  public boolean canPartition(int[] nums) {
  	int total = 0;
    for (int v : nums) total += v;
    
    if (total % 2 != 0) return false;
    
    int half = total / 2;
    final int L = nums.length; 
    boolean[] dp = new boolean[half+1];
    dp[0] = true;
    
    for (int i = 0; i < L; i++) {
      int v = nums[i];
      if (v > half) return false;
      
      for (int sum = half; sum >= v; sum--) {
        dp[sum] = dp[sum] || dp[sum - v];
      }
      if (dp[half]) return true;
    }
    return false;
  }
}

File: Partition_Equal_Subset_Sum/solution1-2.py
class Solution:
    def __init__(self):
      self._dp = {}
  
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        total = sum(nums)
        if total % 2 != 0: return False
        
        half = total // 2
        dp = [False] * (half + 1)
        dp[0] = True
        
        for v in nums:
          if v > half: return False
          
          for target in range(half, v-1, -1):
            dp[target] = dp[target] or dp[target - v]
          
          if dp[half]: return True
        return False
